### PR TITLE
PYTHON-4654 Clean up Async API to match Motor

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

**Comment by ShaneHarvey** [comment]
One test failure:
```
test/asynchronous/test_client.py:1362: error: Value of type variable "_SupportsAcloseT" of "aclosing" cannot be "AsyncMongoClient[Any]"  [type-var]
```

```
 [2024/08/12 15:57:52.649] FAILURE: TypeError: 'AsyncDatabase' object is not callable ()
 [2024/08/12 15:57:52.649] self = <test.asynchronous.test_client.TestClient testMethod=test_contextlib>
 [2024/08/12 15:57:52.649]     async def test_contextlib(self):
 [2024/08/12 15:57:52.649]         client = await async_rs_or_single_client()
 [2024/08/12 15:57:52.649]         await client.pymongo_test.drop_collection("test")
 [2024/08/12 15:57:52.649]         await client.pymongo_test.test.insert_one({"foo": "bar"})
 [2024/08/12 15:57:52.649]     
 [2024/08/12 15:57:52.649]         # The socket used for the previous commands has been returned to the
 [2024/08/12 15:57:52.649]         # pool
 [2024/08/12 15:57:52.649]         self.assertEqual(1, len((await async_get_pool(client)).conns))
 [2024/08/12 15:57:52.649]     
 [2024/08/12 15:57:52.649]         # contextlib async support was added in Python 3.10
 [2024/08/12 15:57:52.649]         if _IS_SYNC or sys.version_info >= (3, 10):
 [2024/08/12 15:57:52.649] >           async with contextlib.aclosing(client):
 [2024/08/12 15:57:52.649] test/asynchronous/test_client.py:1362: 
 [2024/08/12 15:57:52.649] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 [2024/08/12 15:57:52.649] self = <contextlib.aclosing object at 0x7f2759001ae0>
 [2024/08/12 15:57:52.649] exc_info = (None, None, None)
 [2024/08/12 15:57:52.649]     async def __aexit__(self, *exc_info):
 [2024/08/12 15:57:52.649] >       await self.thing.aclose()
 [2024/08/12 15:57:52.649] E       TypeError: 'AsyncDatabase' object is not callable
 [2024/08/12 15:57:52.649] /opt/python/3.10/lib/python3.10/contextlib.py:366: TypeError
```

Should we have both close() and aclose()?

**Comment by blink1073** [comment]
> Should we have both close() and aclose()?

Yeah, I think that is fair.

File: gridfs/asynchronous/grid_file.py [hunk 1]
```diff
@@ -1484,14 +1484,15 @@ def __init__(
     _file: Any
     _chunk_iter: Any
 
-    async def open(self) -> None:
+    async def open(self) -> AsyncGridOut:
```
Comment (ShaneHarvey) [inline gridfs/asynchronous/grid_file.py:1487]: Could you explain this one?

Comment (blink1073) [reply gridfs/asynchronous/grid_file.py:1487]: The current `AgnosticGridOut.open()` returns [itself](https://github.com/mongodb/motor/blob/72e68b75690825871642da363ba911531c4cc4b6/motor/motor_gridfs.py#L163), a fact which is used in the tests.  I tried making my own version of open() that wraps `GridOut.open()` but couldn't figure out how to do so.

Comment (ShaneHarvey) [reply gridfs/asynchronous/grid_file.py:1487]: How about this?:
```

    _open = AsyncCommand(attr_name="open")
...
    @coroutine_annotation
    def open(self):
        return self._framework.chain_return_value(self._open(), self.get_io_loop(), self)
```

Comment (blink1073) [reply gridfs/asynchronous/grid_file.py:1487]: That works, thanks!  I'll remove the change here.

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**